### PR TITLE
Make a patch's source an ArraySource which can load its array

### PR DESCRIPTION
**Pull request by d-chambers**
## Description

Follow-up to #1163. The private `PatchSource` record becomes `ArraySource`: a frozen, hashable description of an array stored in a resource, which can load that array without its patch. It is the primitive a lazy patch, and lazily loaded dense coordinates, can later be built on; neither is added here.

**`ArraySource`** (`dascore/core/source.py`) holds `path`, `format`, `version`, `key`, `dims`, `windows` (half-open, positional), `shape` and `dtype`.

- `load()` reads exactly the described array through the format's `read_array` and checks the shape and dtype it got back. `np.asarray(source)` does the same; nothing is cached.
- Slicing (`source[10:20, 5:]`) composes unit-step slices onto the windows and reads nothing. Anything else raises; `narrow` gives the provenance-only source instead.
- `id` is a digest of which array is named (`shape` and `dtype` follow from the rest and stay out). `to_dict`/`from_dict` are JSON-safe, and the dataclass pickles, so another process needs only the source.
- `key` names the array in the resource. Where it is an absolute HDF5 dataset path (`/raw_data`), the loader slices that dataset directly, and so can anything else: `h5py.File(source.path)[source.key][slices]` needs no DASCore reader. Any other key (a DASDAE group name, a Febus zone, `""`) goes to the format's `read_array` as before. There is no second reader method; a dense coordinate is the same call with the coordinate's path. Stored values come back undecoded.
- Nine single-patch HDF5 formats now declare their data's dataset path as the key in `get_metadata` (AI4EPS, APSensing, DASHDF5, GDR, NeubrexDAS, NeubrexRFS, ODH4, Silixa, Uptech) — one line each, `read_array` untouched. Readers which scale, reshape, stitch, or hold several patches (OptoDAS, Terra15, Febus, NetCDF, DASDAE, PRODML) keep their keys. For those nine the public `source_patch_key` changes from `""` to the path, and with it their source-derived `patch_id`; neither has been released.

**Who makes a source loadable.** Readers still set only `key`. `FiberIO.read` and the scan path describe the whole array when they stamp the source, and `read` narrows it to a contiguous selection. Directory formats, and a reader which stamps after selecting for itself (Sintela), keep provenance only. Stamping touches no filesystem.

**What carries it.** The rule is that a loadable source must load exactly the patch's data.

- `select`, a unit-step `isel`/`sel` slice, and contiguous read selections narrow it.
- Metadata-only changes (`update_attrs`) keep it.
- New data, a changed shape, dims or described dtype, a kernel's result, and `PatchMeta.to_patch(data)` keep the path, format, version and key, and drop the rest. A processed patch's summary reports its source as before.

`patch_id` derivation is unchanged: `scripts/differential_check.py --ref origin/dev` compares 900 calls, all identical, ids included.

**Review.** Codex plus five blind reviewers (correctness, test vacuity by hand mutation, blast radius, redundancy, prose). Fixed from it: `PatchMeta.to_patch` carried a loadable source onto other data (found independently by four reviewers); a per-patch `is_dir()` probe while stamping, which raised on `s3://` without credentials (three reviewers), replaced by the reader's declared `input_type`; the non-contiguous read guard had no test; no reader but DASDAE had its source's `load()` checked, so `tests/test_io/test_common_io.py` now checks every reader, whole and selected — which is what catches the Sintela case. Declined: a compatibility alias for `PatchSource` pickles (never released; on `dev` for hours); validating size/mtime at load (the source is a recipe, and the index holds those fields for a caller which wants them).

Checks: full suite 13430 passed, 130 skipped, 2 xfailed; doctests 259 passed; `pre-commit run --all` clean.

## Changelog

- added: `dascore.io.ArraySource` (replacing the unreleased `PatchSource`) describes a stored array and can load, slice, hash and serialize it without its patch; nine single-patch HDF5 formats report their data's dataset path as `source_patch_key`, so the array can be read by path without a DASCore reader.

## Checklist

I have:

- [x] filled in the Changelog section above (see `dascore/docs/contributing/general_guidelines.qmd`).

I have (if applicable):

- [ ] referenced the GitHub issue this PR closes.
- [x] documented the new feature with docstrings and/or appropriate doc page.
- [x] included tests. See [testing guidelines](https://dascore.org/contributing/testing.html).
- [x] added the "ready_for_review" tag once the PR is ready to be reviewed.

## Summary by CodeRabbit

- **New Features**
  - Added richer array provenance for patches, including source location, dimensions, shape, and data type.
  - Enabled reloadable array sources, including direct loading from stored dataset paths.
  - Added source tracking across supported file formats and patch selections.

- **Bug Fixes**
  - Preserved source information when patches are transformed or metadata is updated.
  - Correctly narrowed or detached source information when data changes or selections are incompatible.
  - Improved validation for source shape and data type consistency.

- **Documentation**
  - Updated format-development guidance to reflect the new array source metadata.

### CODE REVIEW

File: dascore/core/source.py [hunk 1]
```diff
@@ -1,16 +1,166 @@
...
+            self,
+            dims=tuple(dims),
+            windows=tuple((0, x) for x in shape),
+            shape=shape,
+            dtype=np.dtype(dtype),
+        )
+
+    def detach(self) -> ArraySource:
+        """Return the provenance alone, for an array this no longer loads."""
+        return replace(self, dims=(), windows=(), shape=(), dtype=None)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/source.py:105]: Moot as of a6c94b44: the `address` field and `read_address` are gone. A stored array is now named by `key` (an absolute HDF5 path) and read through `read_array`. `detach()` keeps `key` on purpose, since it is provenance, and only the framework calls `describe()`, on sources it has just stamped.

File: dascore/core/source.py [hunk 2]
```diff
@@ -1,16 +1,166 @@
...
+
+    def __array__(self, dtype=None, copy=None) -> np.ndarray:
+        """Load the array for numpy; each call reads the resource."""
+        out = self.load()
+        return out if dtype is None else out.astype(dtype, copy=False)
+
+    def to_dict(self) -> dict[str, Any]:
+        """Return a JSON-compatible dict which `from_dict` reads back."""
+        out = asdict(self)
+        out["dtype"] = None if self.dtype is None else np.dtype(self.dtype).str
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/source.py:149]: Declining. Every reader in the tree describes a plain numeric array, and nothing here can produce a structured dtype, so the lossless form would be unreachable code. If a format with record arrays arrives, `np.lib.format.dtype_to_descr` is the drop-in.

File: dascore/io/core.py [hunk 3]
```diff
@@ -1155,11 +1214,13 @@ def _stamp_source_ids(
...
         patch = patches[index]
         origin = replace(
-            patch._source or PatchSource(),
+            patch._source or ArraySource(),
             path=reload_path,
             format=file_format,
             version=file_version,
         )
+        if describe and reload_path:
+            origin = origin.describe(patch.shape, patch.dtype, patch.dims)
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/io/core.py:1204]: Declining. I tried it: making the path absolute here also changes the public `source_path` that `dc.scan` and `patch.summary` report, which on dev echoes the spelling the caller passed (`a.h5` in, `a.h5` out). A spool already hands out absolute paths from its index, which is the route a source takes to another process; a relative `dc.read` followed by a `chdir` fails loudly on the missing file, or on the shape/dtype check.

File: dascore/proc/basic.py [hunk 4]
```diff
@@ -294,7 +294,11 @@ def update(
     # Each kind keeps what it is: `drop_data` and `to_patch`, not a
     # keyword here, are how data come and go.
     out = self._new_like(data, coords, attrs, dtype)
-    out._source = self._source if source is None else source
+    if source is None and (source := self._source) is not None:
+        # A source loads the array it was made for, not one replaced or relaid.
+        if data is not None or (out.dims, out.shape) != (self.dims, self.shape):
+            source = source.detach()
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/proc/basic.py:301]: Fixed: `update` now detaches when the described dtype changes, alongside dims and shape. Covered by `TestCarry::test_meta_dtype`.

File: dascore/core/source.py [hunk 5]
```diff
@@ -1,16 +1,166 @@
...
-from dataclasses import dataclass
+from dataclasses import asdict, dataclass, replace
+from math import prod
+from typing import Any
+
+import numpy as np
+
+from dascore.exceptions import ParameterError
+from dascore.utils.serialize import digest
+
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply dascore/core/source.py:14]: Moot as of a6c94b44: there is no `address` any more. `key` is the one name for an array in a resource and is always what `read_array` is given, so it belongs in the id.

File: tests/test_core/test_source.py [hunk 6]
```diff
@@ -0,0 +1,248 @@
...
+    def test_too_many_indices(self, source):
+        """More indices than dimensions is an error."""
+        with pytest.raises(IndexError):
+            source[:, :, :]
+
+
+class TestIdentity:
+    """Equal sources hash alike; a different array does not."""
+
+    def test_hash_and_id(self, source):
```
[the remark this answers is not in the document]
Comment (d-chambers) [reply tests/test_core/test_source.py:141]: Done: `test_hash_and_id` now pins the id of a fixed synthetic source.

**Comment by d-chambers** [comment]
@coderabbitai re the out-of-diff comment on `coords.py:769`: accepted. `Isel.get_metadata` now narrows the source as `Select` does, so a unit-step `isel`/`sel` slice stays loadable and scalar, strided, reversed and array indexers detach. Covered by `TestCarry::test_isel_and_sel`.
